Removed name of slots and changed projection to produce LongSlots

SaschaPeukert opened the pull request.

SaschaPeukert committed:
Removed name of slots and changed projection to produce LongSlots

SaschaPeukert committed:
Removed left-over occurrences of the removed name property for slots

henriknyman reviewed (dismissed):
Removing name from slots makes for a horrible debugging experience

systay commented:
We removed the name from slot because we now can have two different variables sharing the same slot. The name information is still available in the `PipelineInformation`.

Commit:
Make sure to clone the mutable state when breaking a pipeline

Commit:
Rename method to make it easier to understand

henriknyman commented:
@systay I understand it was because of the multiple aliased names, but I still think we need to at least project this information in the implementation of  `PrimitiveExecutionContext.iterator` (that is used for debugging) or I strongly believe it will slow down development too much. Also currently there is no backward map from slots to names so the information is not easily available. 

We could either keep the name in the slot as a Set or Seq or add a separate map from slot to set of names. (Even keeping the first name would be better than not fixing this since it would work in >95% of cases).

henriknyman committed:
Improve debug information of PrimitiveExecutionContext

- Save aliases of PipelineInformation in a multi-map
from slots to names
- Preserve aliases when cloning pipeline

henriknyman commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/PipelineInformation.scala:
```diff
@@ -134,38 +139,51 @@ object PipelineInformation {
-  def seedClone(): PipelineInformation = {
-    new PipelineInformation(this.slots, numberOfLongs, numberOfReferences)
+  def breakPipelineAndClone(): PipelineInformation = {
+    new PipelineInformation(this.slots.clone(), numberOfLongs, numberOfReferences)
```
We are losing alias information here

henriknyman committed:
Fix ORDER BY comparator for nullable primitives

henriknyman committed:
Adapt UnionSlottedPipe to unnamed slots

- Fix bug in computeUnionMapping for incoming longslots
and outgoing refslots
- Deduplicate code for project slot expression
- Pre-compute project slot expression

henriknyman committed:
Fixup debug info

henriknyman committed:
Fix bug in slotted CartesianProduct

Since CartesianProductSlottedPipe copies slots using array copy,
the slot ordering must be preserved when adding new slots to the
outgoing pipeline.

henriknyman commented:
We still need to deduplicate aliased slots in some places. I have a WIP-commit for that which I will try to finish off. After that somebody else should review again since there are a lot of changes.

systay reviewed (changes requested):
Nice changes! Could you please add a few tests to make sure the bugs you have fixed don't reappear in the future?
